### PR TITLE
implement s3 storage

### LINKED ISSUE #69: Implement S3 for Storage
Implement an S3 interface for [Storage](https://github.com/scratchdata/ScratchDB/blob/v2/pkg/filestore/storage.go). 

- The settings should be configurable via `config.go`
- We should instantiate this object in the main function if the s3 interface is present

### PR DESCRIPTION
**Pull request by sd1li**
Close #69

### CODE REVIEW

**Comment by sd1li** [comment]
@poundifdef This PR is waiting on (and includes) the changes inhttps://github.com/scratchdata/ScratchDB/pull/80, but other than that, it's ready for review.

I added the ability to automatically create the bucket if it's missing. I'm not sure if that's desirable. If not, let me know and I can delete it.

File: pkg/filestore/s3/s3.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	var awsErr awserr.Error
+	if errors.As(err, &awsErr) && awsErr.Code() == s3.ErrCodeNoSuchKey {
+		return fmt.Errorf("Storage.Upload: %s: %w", path, filestore.ErrNotFound)
+	}
+	return fmt.Errorf("Storage.Download: %s: %w", path, err)
+}
+
+// NewStorage returns a new initialized Storage
+func NewStorage(ctx context.Context, c config.S3) *Storage {
+	awsCreds := credentials.NewStaticCredentials(c.AccessKeyId, c.SecretAccessKey, "")
```
Comment (poundifdef) [inline pkg/filestore/s3/s3.go:84]: This is a duplicate of line 93.

**Review by poundifdef — CHANGES_REQUESTED** [review]

**Comment by poundifdef** [comment]
Does this work when you run the code locally? The test would be to make a POST to the /data endpoint and see if data is successfully uploaded to S3.

File: pkg/filestore/s3/s3.go (hunk 1 above)
Comment (sd1li) [reply pkg/filestore/s3/s3.go]: done

File: pkg/filestore/s3/s3.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	_, err = s.client.PutObjectWithContext(s.ctx, input)
+	if err == nil {
+		return nil
+	}
+	var awsErr awserr.Error
+	if !errors.As(err, &awsErr) || awsErr.Code() != s3.ErrCodeNoSuchBucket {
+		return fmt.Errorf("Storage.Upload: %s: %w", path, err)
+	}
+
+	// if the bucket doesn't exist, create it and try again
```
Comment (poundifdef) [inline pkg/filestore/s3/s3.go:52]: I like the thought, but let's omit auto-creating buckets for now.

**Comment by sd1li** [comment]
Sending requests to `/data` hangs, I assume because `queuestorage` (on the `v2` branch) is incomplete.

File: pkg/filestore/s3/s3.go (hunk 2 above)
Comment (sd1li) [reply pkg/filestore/s3/s3.go]: done

**Review by chumaumenze — CHANGES_REQUESTED** [review]

File: pkg/filestore/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+// Storage implements filestore.StorageBackend using Amazon S3
+type Storage struct {
+	ctx        context.Context
```
[the remark this answers is not in the document]
Comment (sd1li) [reply pkg/filestore/s3/s3.go]: done

**Review by chumaumenze — APPROVED** [review]

**Comment by chumaumenze** [comment]
@sd1li LGTM. Resolve conflicts.

**Review by poundifdef — APPROVED** [review]

**Pull request merged**